modules/python-modules/README.md

```diff
@@ -0,0 +1,648 @@
+        return True
+
+    def message_acked(self, acked_message_bookmark):
+        pass
+```
+
+While ConsecutiveAckTracker() seems to provide a much more useful service,
+but InstantAckTracker() performs better, as it does not have to track
+acknowledgements of individual messages.
+
+The most complex scenario is implemented by BatchedAckTracker, this allows
+you to track the acks for individual messages, as they happen, not enforcing
+any kind of ordering.
+
+```python
+
+class MyFetcher(LogFetcher):
+
+    ...
+
+    def init(self, options):
+        self.ack_tracker = BatchedAckTracker(timeout=500, batch_size=100,
+                                             batched_ack_callback=self.messages_acked)
+        return True
+
+    def messages_acked(self, acked_message_bookmarks):
+        pass
+```
+
+BatchedAckTracker would call your callback every now and then, as specified
+by the `timeout` argument in milliseconds. `batch_size` specifies the number of
+outstanding messages at a time. With this interface it's quite easy to
+perform acknowledgements back to the source interface where per-message
+acknowledgements are needed (e.g. Google PubSub).
+
+### Creating a syslog-ng a LogSource based Source plugin
+
+While LogFetcher gives us a convinient interface for fetching messages from
+backend services via blocking APIs, it is limited to performing the fetching
+operation in a sequential manner: you fetch a batch of messages, feed the
+syslog-ng pipeline with those messages and then repeat.
+
+LogSource is more low-level but allows the use of an asynchronous framework
+(asyncio, etc) to perform message fetching along multiple threads of
+execution.
+
+This sample uses asyncio to generate two independent sequences of messages,
+one is generated every second, the other is every 1.5 seconds, running
+concurrently via an asyncio event loop.
+
+It is also pretty easy to create a source that implements a HTTP server,
+and which injects messages coming via HTTP to the syslog-ng pipeline.
+
+```python
+
+from syslogng import LogSource
+from syslogng import LogMessage
+import asyncio
+
+class MySource(LogSource):
+    cancelled = False
+
+    def run(self):
+        asyncio.run(self.main())
+
+    async def main(self):
+        await asyncio.gather(self.sequence1(), self.sequence2())
+
+    async def sequence1(self):
+        while not self.cancelled:
+            await asyncio.sleep(1)
+            self.post_message(LogMessage("message1"))
+
+    async def sequence2(self):
+        while not self.cancelled:
+            await asyncio.sleep(1.5)
+            self.post_message(LogMessage("message2"))
+
+    def request_exit(self):
+        self.cancelled = True
+
+```
+
+Acknowledgement mechanisms (ConsecutiveAckTracker, BatchedAckTracker) can be
+used similarly to how it was described at LogFetcher.
+
+## Making it more native
+
+All the examples above used some form of python() driver to be actually
```

Comment:
[optional]

python() can be considered as source code, so it could be backticked.
```suggestion
All the examples above used some form of `python()` driver to be actually
```